sql/LabelCenterline.sql

```diff
@@ -0,0 +1,71 @@
+/******************************************************************************
+### LabelCenterline ###
+
+Given a polygon or multipolygon, calculates a linestring suitable for a label.
+
+__Parameters:__
+
+- `geometry` inGeometry - A polygon or multipolygon.
+
+__Returns:__ `geometry(multiline)`
+******************************************************************************/
+CREATE OR REPLACE FUNCTION CountDisconnectedEndpoints(polyline geometry, testline geometry)
+    RETURNS integer AS $$
+    BEGIN
+        RETURN ST_NPoints(ST_RemoveRepeatedPoints(ST_Points(polyline)))
+                    - ST_NPoints(ST_RemoveRepeatedPoints(ST_Points(ST_Difference(polyline, testline))))
+                    - ST_NPoints(testline) + 2;
+    END
+    $$ LANGUAGE plpgsql;
+CREATE OR REPLACE FUNCTION TrimmedCenterline(inPolyline geometry)
+    RETURNS geometry AS $$
+    DECLARE outPolyline geometry;
+    BEGIN
+        WITH tbla AS (
+            SELECT inPolyline as polyline, (ST_Dump(inPolyline)).geom as edge
+        ),
+        tblb AS (
+            SELECT polyline, edge as shortestBranchLine
+            FROM tbla
+            WHERE CountDisconnectedEndpoints(polyline, edge) > 0
+            ORDER BY ST_Length(edge) ASC
+            LIMIT 1
+        ),
+        tblc AS (
+            SELECT ST_LineMerge(ST_Difference(polyline, shortestBranchLine)) as polyline
+            FROM tblb
+        )
+        SELECT TrimmedCenterline(polyline) as polyline
+        FROM tblc
+        WHERE ST_NumGeometries(polyline) > 1
+        UNION ALL
+        SELECT polyline INTO outPolyline FROM tblc;
+        RETURN outPolyline;
+    END
+    $$ LANGUAGE plpgsql;
+CREATE OR REPLACE FUNCTION LabelCenterline(inGeometry geometry)
+    RETURNS geometry AS $$
+    DECLARE outPolyline geometry;
+    BEGIN
+        WITH tbla AS (
+            SELECT inGeometry as inPolygon WHERE ST_GeometryType(inGeometry) = 'ST_Polygon'
+            UNION ALL
+            SELECT ST_ConcaveHull(ST_Simplify(inGeometry, 25), 0.2) as inPolygon WHERE ST_GeometryType(inGeometry)='ST_MultiPolygon'
```

Comment:
I'm skeptical about this approach for multipolygons, as it can lead to geoms outside the original area. This might not be an issue for water, but it is for other types of polygons. If the function is suitable for water only, then it should be renamed.

Reply:
I see your point. I was running into some errors with ST_ConcaveHull anyway, so I just made it calculate a separate centerline for each component polygon. Future work would be to add a parameter giving more control over what to do with multipolygons. There are cases where you'd want one combined centerline, like for an archipelago.

<img width="589" alt="Screenshot 2023-03-05 at 2 13 38 PM" src="https://user-images.githubusercontent.com/2046746/222981835-a2d27e4a-e9f3-48f4-9d0c-f4ca3558103f.png">